File: kernel/core/reconfs_check.c
/* ReconFS -- the checker, which is two derivations that must agree.
 *
 * --- Why this exists at all ---
 *
 * Nobody else implements ReconFS. For partition tables the reader could be
 * checked against disks written by sgdisk and sfdisk, which share no code with
 * this and no misunderstanding with it. Here there is no second implementation,
 * and testing a reader against its own writer is the failure mode
 * scripts/make-partition-fixtures.sh opens by naming: a writer and a reader
 * built from the same misunderstanding agree perfectly.
 *
 * So the independence is built into the format instead. Two derivations of one
 * fact -- which object owns which block -- from disjoint fields:
 *
 *   forward   from the root, follow directory entries and block pointers down,
 *             reading pointers and nothing else
 *   reverse   read the owner table, and nothing else
 *
 * Neither reads a field the other reads. If they agree, that agreement is
 * evidence. If they had shared a field they would agree by construction, which
 * is worth nothing.
 *
 * --- The rule this file obeys ---
 *
 * The reverse sweep calls nothing the reader uses to resolve a path. It reads
 * owner-table leaves and nothing else. A sweep that quietly used the reader's
 * lookup would be the reader checking itself with extra steps.
 */
#include <recon/kernel/reconfs.h>

#include <recon/kernel/block.h>
#include <recon/kernel/console.h>
#include <recon/kernel/heap.h>
#include <recon/kernel/kstring.h>

#define INODE_CSUM	RK_OFFSETOF(struct reconfs_inode, checksum)

/* The forward walk keeps one dossier number per block on the volume, so the two
 * derivations can be compared entry by entry rather than as two sets.
 *
 * That costs eight bytes a block, which is what caps the volume this can check
 * in one pass. A larger volume needs the check done in segments -- reading the
 * table once per segment and walking the tree once per segment -- which is a
 * real piece of work and is not written. The cap is enforced rather than
 * exceeded, because a checker that quietly examines part of a volume and
 * reports no disagreements is the failure this whole file exists to avoid. */
#define CHECK_MAX_BLOCKS	(1ULL << 22)	/* 16GiB at a 4KiB block */

struct walk {
	struct reconfs *fs;
	u64 *by;		/* by[block] = the dossier the tree says owns it */
	struct reconfs_check_result *r;
	u32 depth;
};

static void disagree(struct reconfs_check_result *r, const char *why, u64 blk)
{
	if (!r->disagreements) {
		r->first_disagreement = why;
		r->first_disagreement_block = blk;
	}
	r->disagreements++;
}

/* Records that `owner` reaches `blk`. A block reached twice is a corruption,
 * not a shortcut to skip. */
static void claim(struct walk *w, u64 blk, u64 owner, const char *what)
{
	if (blk < reconfs_super_b(w->fs->block_size) ||
	    blk >= w->fs->total_blocks) {
		disagree(w->r, "a pointer left the volume", blk);
		return;
	}

	if (w->by[blk]) {
		disagree(w->r, what, blk);
		return;
	}

	w->by[blk] = owner;
	w->r->blocks_seen_forward++;
}

static enum reconfs_status walk_file(struct walk *w,
				     const struct reconfs_inode *ino)
{
	unsigned i;

	for (i = 0; i < RECONFS_DIRECT; i++)
		if (ino->direct[i])
			claim(w, ino->direct[i], ino->dossier,
			      "a data block is reachable twice");

	if (ino->indirect) {
		u64 *slots = kzalloc(w->fs->block_size);
		enum reconfs_status st;
		unsigned j;

		if (!slots)
			return RECONFS_ERR_NOMEM;

		claim(w, ino->indirect, ino->dossier,
		      "an indirect block is reachable twice");

		st = reconfs_read_block(w->fs, ino->indirect, slots,
					RECONFS_NO_CSUM);
		if (st == RECONFS_OK) {
			for (j = 0; j < w->fs->per_index; j++)
				if (slots[j])
					claim(w, slots[j], ino->dossier,
					      "a data block is reachable twice");
		}
		kfree(slots);
		if (st != RECONFS_OK)
			return st;
	}

	/* Part of the format, not written by this version. A file that used one
	 * is reported rather than skipped: a checker that silently ignores a
	 * field will agree with a writer that silently corrupts it. */
	if (ino->double_indirect || ino->triple_indirect)
		disagree(w->r, "a file uses indirection this version cannot write", 0);

	return RECONFS_OK;
}

static enum reconfs_status walk_inode(struct walk *w, u64 blk,
				      u64 expect_parent, u64 owner);

static enum reconfs_status walk_dir_block(struct walk *w, u64 dir_blk,
					  u64 dir_dossier, const u8 *buf,
					  u32 len)
{
	u32 off = 0;

	while (off + RECONFS_DIRENT_MIN <= len) {
		const struct reconfs_dirent *de =
			(const struct reconfs_dirent *)(buf + off);
		enum reconfs_status st;

		if (!de->rec_len)
			break;			/* padding to the block end */

		if (de->rec_len < RECONFS_DIRENT_MIN ||
		    off + de->rec_len > len ||
		    (u32)de->name_len + RECONFS_DIRENT_MIN > de->rec_len) {
			disagree(w->r, "a directory entry does not fit its block",
				 dir_blk);
			return RECONFS_ERR_CORRUPT;
		}

		if (de->inode) {
			st = walk_inode(w, de->inode, dir_dossier, dir_dossier);
			if (st != RECONFS_OK)
				return st;
		}

		off += de->rec_len;
	}

	return RECONFS_OK;
}

/* `owner` is the dossier that should be recorded against this inode's own
 * block -- its parent's, or the archive for the root. */
static enum reconfs_status walk_inode(struct walk *w, u64 blk,
				      u64 expect_parent, u64 owner)
{
	struct reconfs_inode *ino;
	enum reconfs_status st;

	/* Depth, not a visited set: a cycle shows up as a block reached twice,
	 * which `claim` already reports. This is only here so a cycle cannot run
	 * the stack out before that happens. */
	if (w->depth > 64) {
		disagree(w->r, "the directory tree is deeper than 64", blk);
		return RECONFS_ERR_CORRUPT;
	}

	claim(w, blk, owner, "an inode is reachable by two paths");

	ino = kzalloc(w->fs->block_size);
	if (!ino)
		return RECONFS_ERR_NOMEM;

	st = reconfs_read_block(w->fs, blk, ino, INODE_CSUM);
	if (st != RECONFS_OK) {
		disagree(w->r, "an inode did not read", blk);
		kfree(ino);
		return st;
	}

	if (ino->magic != RECONFS_INODE_MAGIC) {
		disagree(w->r, "a directory entry points at something that is "
			       "not an inode", blk);
		kfree(ino);
		return RECONFS_ERR_CORRUPT;
	}

	if (!ino->dossier) {
		disagree(w->r, "an inode has no dossier number", blk);
		kfree(ino);
		return RECONFS_ERR_CORRUPT;
	}

	/* The forward walk knows who the parent is because it came from there.
	 * The inode also says. They must match -- the one place the two
	 * derivations are allowed to meet, and it is a comparison rather than a
	 * read: nothing downstream depends on `parent` being right. */
	if (ino->parent != expect_parent)
		disagree(w->r, "an inode's parent is not the directory that "
			       "names it", blk);

	w->r->inodes++;

	if (ino->type == RECONFS_TYPE_DIR) {
		unsigned i;

		w->depth++;
		if (ino->inline_len) {
			st = walk_dir_block(w, blk, ino->dossier, ino->data,
					    ino->inline_len);
		} else {
			u8 *db = kzalloc(w->fs->block_size);

			if (!db) {
				st = RECONFS_ERR_NOMEM;
			} else {
				for (i = 0; i < RECONFS_DIRECT; i++) {
					if (!ino->direct[i])
						continue;
					claim(w, ino->direct[i], ino->dossier,
					      "a directory block is reachable twice");
					st = reconfs_read_block(w->fs,
								ino->direct[i],
								db,
								RECONFS_NO_CSUM);
					if (st != RECONFS_OK)
						break;
					st = walk_dir_block(w, blk,
							    ino->dossier, db,
							    w->fs->block_size);
					if (st != RECONFS_OK)
						break;
				}
				kfree(db);
			}
		}
		w->depth--;
	} else if (ino->type == RECONFS_TYPE_FILE) {
		if (!ino->inline_len)
			st = walk_file(w, ino);
	} else {
		disagree(w->r, "an inode has a type this format does not define",
			 blk);
		st = RECONFS_ERR_CORRUPT;
	}

	kfree(ino);
	return st;
}

/* Claims the blocks the owner table itself occupies: its index blocks and its
 * leaves, from the root down.
 *
 * This is the forward walk's business, not the sweep's -- it reads the table's
 * *structure*, following pointers, which is what the forward walk does. The
 * sweep still reads only the table's contents. The two stay disjoint.
 */
static enum reconfs_status claim_table(struct walk *w, u64 node, u32 depth)
{
	u64 *slots;
	enum reconfs_status st;
	u32 i;

	if (node >= w->fs->total_blocks) {
		disagree(w->r, "the owner table leaves the volume", node);
		return RECONFS_ERR_CORRUPT;
	}

	if (w->by[node]) {
		disagree(w->r, "an owner table block is reachable twice", node);
		return RECONFS_ERR_CORRUPT;
	}
	w->by[node] = RECONFS_OWNER_ARCHIVE;
	w->r->blocks_seen_forward++;

	if (depth == 0)
		return RECONFS_OK;

	slots = kzalloc(w->fs->block_size);
	if (!slots)
		return RECONFS_ERR_NOMEM;

	st = reconfs_read_block(w->fs, node, slots, RECONFS_NO_CSUM);
	if (st == RECONFS_OK) {
		for (i = 0; i < w->fs->per_index; i++) {
			if (!slots[i])
				continue;
			st = claim_table(w, slots[i], depth - 1);
			if (st != RECONFS_OK)
				break;
		}
	}

	kfree(slots);
	return st;
}

/* --- The reverse sweep, and the comparison --------------------------------
 *
 * Reads the owner table, a whole leaf at a time, and compares each entry with
 * what the forward walk recorded. Nothing here follows a directory entry, reads
 * an inode, or resolves a name.
 */
static enum reconfs_status sweep_and_compare(struct reconfs *fs,
					     const u64 *by,
					     struct reconfs_check_result *r)
{
	u64 *leaf = kzalloc(fs->block_size);
	u64 b;

	if (!leaf)
		return RECONFS_ERR_NOMEM;

	for (b = 0; b < fs->total_blocks; b++) {
		u64 actual, expected;

		/* A whole leaf at a time. Asking for one owner at a time walks
		 * the tree from the root on every block, which on a volume of
		 * any size is the difference between a check that is slow and
		 * one that never finishes. */
		if (b % fs->per_leaf == 0) {
			enum reconfs_status st;

			st = reconfs_read_leaf(fs, b / fs->per_leaf, leaf);
			if (st != RECONFS_OK) {
				kfree(leaf);
				return st;
			}
		}

		actual = leaf[b % fs->per_leaf];
		expected = by[b];

		if (actual != RECONFS_OWNER_VOID)
			r->blocks_seen_reverse++;

		if (actual == expected)
			continue;

		if (actual == RECONFS_OWNER_VOID)
			disagree(r, "reachable from the root but not allocated", b);
		else if (!expected)
			disagree(r, "allocated but not reachable from the root", b);
		else
			disagree(r, "allocated to one object and reachable from "
				    "another", b);
	}

	kfree(leaf);
	return RECONFS_OK;
}

enum reconfs_status reconfs_check(struct reconfs *fs,
				  struct reconfs_check_result *out)
{
	struct walk w;
	enum reconfs_status st;

	kmemset(out, 0, sizeof(*out));
	kmemset(&w, 0, sizeof(w));

	if (!fs->mounted)
		return RECONFS_ERR_NOT_RECONFS;

	if (fs->total_blocks > CHECK_MAX_BLOCKS) {
		kputs("reconfs: volume too large to check in one pass with the "
		      "memory available\n");
		return RECONFS_ERR_NOMEM;
	}

	w.fs = fs;
	w.r  = out;
	w.by = kzalloc((size_t)(fs->total_blocks * sizeof(u64)));
	if (!w.by)
		return RECONFS_ERR_NOMEM;

	/* --- The archive's own blocks, accounted for like everything else ---
	 *
	 * The comparison used to *exempt* every block owned by the archive from
	 * needing to be reachable, on the grounds that the superblocks and the
	 * owner table are owned by nothing above them. That is true of those
	 * blocks and of nothing else -- and the exemption applied to any block
	 * whose owner happened to be the archive.
	 *
	 * Which included every stale copy of the root directory. Each commit
	 * writes a new root inode, and nothing released the old one (KF-122):
	 * a leaked block on every create, rename, write and remove, owned by
	 * the archive because the root has no parent, and therefore skipped by
	 * the one check that would have named it.
	 *
	 * So the archive's blocks are claimed here, explicitly, and the
	 * exemption is gone. Every allocated block on the volume must be
	 * reachable from something.
	 */
	{
		u64 b, first = reconfs_super_b(fs->block_size) + 1;

		/* Both superblocks and the run reserved between them. */
		for (b = 0; b < first; b++) {
			if (b >= fs->total_blocks)
				break;
			w.by[b] = RECONFS_OWNER_ARCHIVE;
			out->blocks_seen_forward++;
		}

		st = claim_table(&w, fs->table_root, fs->table_depth);
		if (st != RECONFS_OK) {
			kfree(w.by);
			return st;
		}
	}

	/* The root's own block is archive-owned: it has no parent, and that is
	 * what "no parent" looks like in the table. */
	st = walk_inode(&w, fs->root_inode, 0, RECONFS_OWNER_ARCHIVE);
	if (st == RECONFS_OK)
		st = sweep_and_compare(fs, w.by, out);

	kfree(w.by);
	return st;
}

File: scripts/reconfs-check.py
#!/usr/bin/env python3
"""Read a ReconFS image and say whether it holds together.

--- Why this exists in a different language ---

Nobody else implements ReconFS. Testing the kernel's checker against the
kernel's writer proves only that they share their assumptions, which is the
failure mode scripts/make-partition-fixtures.sh opens by naming: a writer and a
reader built from the same misunderstanding agree perfectly.

So this is written from kernel/include/recon/kernel/reconfs.h -- from the field
offsets and the rules, not from kernel/core/reconfs*.c. Where the C and this
disagree, one of them is wrong, and neither was derived from the other.

It is also what judges a power cut. The kernel cannot check an image it was
killed in the middle of writing; something outside has to read what survived.

--- What it checks ---

  * a superblock validates, and the one with the higher epoch is believed
  * every block's checksum, where the block kind carries one
  * the forward walk: root -> directory entries -> inodes -> block pointers
  * the reverse sweep: every owner-table entry
  * the two agree, block for block

It writes nothing, ever. That is what makes it safe to run on an image that is
about to be examined again.
"""
import struct
import sys

BLOCK_MIN = 4096
BLOCK_MAX = 65536

MAGIC       = 0x3153466E6F636552      # "ReconFS1"
INODE_MAGIC = 0x316F6E496E636552      # "ReconIno1"
VERSION     = 1

OWNER_VOID    = 0
OWNER_ARCHIVE = 1
DOSSIER_ROOT  = 2

TYPE_FILE = 1
TYPE_DIR  = 2

DIRECT = 12

# --- Layouts, by explicit offset -------------------------------------------
#
# Field by field rather than as one struct format string. A format string is a
# single opaque token where one wrong letter shifts everything after it and the
# values still parse -- which is precisely the kind of wrong that looks right.
# These offsets were read out of the compiler with __builtin_offsetof and are
# checked against sizeof below.
SUPER_AT = {
    "magic": ("<Q", 0), "version": ("<I", 8), "block_size": ("<I", 12),
    "epoch": ("<Q", 16), "total_blocks": ("<Q", 24), "root_inode": ("<Q", 32),
    "table_root": ("<Q", 40), "table_depth": ("<I", 48),
    "next_dossier": ("<Q", 56), "blocks_used": ("<Q", 64),
    "fold": ("<I", 152), "seek_is_free": ("<B", 156),
    "discard_supported": ("<B", 157), "transfer_hint": ("<I", 160),
}
SUPER_CSUM = 204
SUPER_SIZE = 208

INODE_AT = {
    "magic": ("<Q", 0), "dossier": ("<Q", 8), "parent": ("<Q", 16),
    "type": ("<I", 24), "mode": ("<I", 28), "uid": ("<I", 32),
    "gid": ("<I", 36), "size": ("<Q", 40), "btime": ("<Q", 48),
    "mtime": ("<Q", 56), "ctime": ("<Q", 64), "links": ("<I", 72),
    "flags": ("<I", 76), "inline_len": ("<I", 80),
    "indirect": ("<Q", 184), "double": ("<Q", 192), "triple": ("<Q", 200),
}
INODE_DIRECT_AT = 88
INODE_CSUM = 224
INODE_SIZE = 232          # where the inline data begins

DIRENT_SIZE = 16          # inode, rec_len, name_len, type, reserved


def field(buf, table, name):
    fmt, off = table[name]
    return struct.unpack_from(fmt, buf, off)[0]


def crc32_reflected(data: bytes) -> int:
    """The same reflected CRC-32 the format uses. Bitwise on purpose: this is a
    checker, and a table would be one more thing to have copied wrong."""
    crc = 0xFFFFFFFF
    for byte in data:
        crc ^= byte
        for _ in range(8):
            crc = (crc >> 1) ^ (0xEDB88320 if crc & 1 else 0)
    return crc ^ 0xFFFFFFFF


class Bad(Exception):
    pass


class Image:
    def __init__(self, path):
        with open(path, "rb") as f:
            self.data = f.read()
        self.bs = None
        self.super = None
        self.live = None

    def raw(self, offset, length):
        if offset + length > len(self.data):
            raise Bad(f"read past the end of the image at {offset}")
        return self.data[offset:offset + length]

    def block(self, n):
        return self.raw(n * self.bs, self.bs)

    # --- Superblocks -------------------------------------------------------
    def parse_super(self, buf):
        return {k: field(buf, SUPER_AT, k) for k in SUPER_AT}

    def load(self):
        """Both superblocks sit at fixed byte offsets -- 0 and BLOCK_MAX -- so
        either can be found without knowing the block size the other records."""
        best = None
        for at in (0, BLOCK_MAX):
            try:
                head = self.raw(at, BLOCK_MIN)
            except Bad:
                continue

            s = self.parse_super(head)
            if s["magic"] != MAGIC or s["version"] != VERSION:
                continue

            bs = s["block_size"]
            if bs < BLOCK_MIN or bs > BLOCK_MAX or (bs & (bs - 1)):
                continue

            full = self.raw(at, bs)
            off = SUPER_CSUM
            stored = struct.unpack("<I", full[off:off + 4])[0]
            body = full[:off] + b"\0\0\0\0" + full[off + 4:]
            if crc32_reflected(body) != stored:
                continue

            if best is None or s["epoch"] > best[0]["epoch"]:
                best = (s, at, bs)

        if best is None:
            raise Bad("no valid superblock")

        self.super, self.live, self.bs = best
        return self.super

    # --- The owner table ---------------------------------------------------
    def per_leaf(self):
        return self.bs // 8

    def per_index(self):
        return self.bs // 8

    def leaf_block(self, leaf_index):
        node = self.super["table_root"]
        depth = self.super["table_depth"]
        span = self.per_index() ** depth
        idx = leaf_index

        while depth:
            slots = struct.unpack(f"<{self.per_index()}Q", self.block(node))
            span //= self.per_index()
            slot = idx // span
            if slot >= self.per_index():
                raise Bad("owner table index out of range")
            node = slots[slot]
            idx %= span
            if not node:
                raise Bad("owner table has a hole")
            depth -= 1

        return node

    def owners(self):
        """Every block's owner, in order. Reads leaves, and nothing else."""
        total = self.super["total_blocks"]
        out = []
        for first in range(0, total, self.per_leaf()):
            leaf = self.block(self.leaf_block(first // self.per_leaf()))
            vals = struct.unpack(f"<{self.per_leaf()}Q", leaf)
            take = min(self.per_leaf(), total - first)
            out.extend(vals[:take])
        return out

    # --- Inodes and names --------------------------------------------------
    def inode(self, blk):
        buf = self.block(blk)
        i = {k: field(buf, INODE_AT, k) for k in INODE_AT}
        i["direct"] = list(struct.unpack_from("<12Q", buf, INODE_DIRECT_AT))

        if i["magic"] != INODE_MAGIC:
            raise Bad(f"block {blk} is not an inode")

        stored = struct.unpack_from("<I", buf, INODE_CSUM)[0]
        body = buf[:INODE_CSUM] + b"\0\0\0\0" + buf[INODE_CSUM + 4:]
        if crc32_reflected(body) != stored:
            raise Bad(f"inode at block {blk} does not match its checksum")

        i["_raw"] = buf
        return i

    def contents(self, ino):
        """A file's bytes, assembled from wherever they live."""
        want = ino["size"]

        if ino["inline_len"]:
            return ino["_raw"][INODE_SIZE:INODE_SIZE + want]

        out = bytearray()
        n = 0
        slots = None
        while len(out) < want:
            if n < 12:
                where = ino["direct"][n]
            else:
                if slots is None:
                    if not ino["indirect"]:
                        raise Bad("a file needs an indirect block and has none")
                    slots = struct.unpack(f"<{self.per_index()}Q",
                                          self.block(ino["indirect"]))
                if n - 12 >= len(slots):
                    raise Bad("a file is longer than its pointers reach")
                where = slots[n - 12]
            if not where:
                raise Bad("a file has a hole in it")
            take = min(self.bs, want - len(out))
            out += self.block(where)[:take]
            n += 1

        return bytes(out)

    def dir_entries(self, ino):
        """Every (name, child, type) in a directory, in stored order."""
        streams = []
        if ino["inline_len"]:
            streams.append(ino["_raw"][INODE_SIZE:
                                       INODE_SIZE + ino["inline_len"]])
        else:
            for b in ino["direct"]:
                if b:
                    streams.append(self.block(b))

        out = []
        for s in streams:
            off = 0
            while off + DIRENT_SIZE <= len(s):
                child, rec_len, name_len, etype = struct.unpack(
                    "<QHBB", s[off:off + 12])
                if rec_len == 0:
                    break
                if rec_len < DIRENT_SIZE or off + rec_len > len(s) or \
                        DIRENT_SIZE + name_len > rec_len:
                    raise Bad("a directory entry does not fit its block")
                name = s[off + DIRENT_SIZE:off + DIRENT_SIZE + name_len].decode(
                    "utf-8", "replace")
                if child:
                    out.append((name, child, etype))
                off += rec_len
        return out


PAYLOAD_MAGIC = 0x52464350          # "RFCP"


def reconfs_super_b_blocks(img):
    """The first block a file may use: one past the second superblock."""
    return BLOCK_MAX // img.bs + 1



def check_payload(data):
    """Is this one complete version of the crash workload's file?

    The round number is at both ends, so a file assembled from the head of one
    version and the tail of another shows two different numbers -- visible
    without knowing which version was supposed to be there, which matters
    because after a power cut nobody does.
    """
    if len(data) < 24:
        return f"only {len(data)} bytes"

    magic, round_a, length, stored = struct.unpack_from("<IIII", data, 0)
    round_b = struct.unpack_from("<I", data, 16)[0]
    round_c = struct.unpack_from("<I", data, len(data) - 4)[0]

    if magic != PAYLOAD_MAGIC:
        return f"magic is {magic:#x}"
    if length != len(data):
        return f"says {length} bytes, is {len(data)}"
    if not (round_a == round_b == round_c):
        return (f"torn: the round is {round_a} at the front, {round_b} after "
                f"the header, {round_c} at the end")

    crc = crc32_reflected(data[:12] + data[16:])
    if crc != stored:
        return f"checksum {crc:08x}, stored {stored:08x}"

    return None


def check(path, want=None):
    img = Image(path)
    sb = img.load()

    total = sb["total_blocks"]
    by = [0] * total          # what the forward walk says owns each block
    problems = []
    inodes = 0

    def claim(blk, owner, why):
        if blk >= total:
            problems.append(f"{why}: block {blk} is past the end")
            return
        if by[blk]:
            problems.append(f"block {blk} is reachable twice")
            return
        by[blk] = owner

    def walk(blk, expect_parent, owner, depth=0):
        nonlocal inodes
        if depth > 64:
            problems.append("the directory tree is deeper than 64")
            return
        claim(blk, owner, "inode")
        ino = img.inode(blk)
        inodes += 1

        if ino["parent"] != expect_parent:
            problems.append(
                f"inode at {blk} says its parent is {ino['parent']}, "
                f"but {expect_parent} names it")

        if ino["type"] == TYPE_DIR:
            if not ino["inline_len"]:
                for b in ino["direct"]:
                    if b:
                        claim(b, ino["dossier"], "directory block")
            for _name, child, _t in img.dir_entries(ino):
                walk(child, ino["dossier"], ino["dossier"], depth + 1)
        elif ino["type"] == TYPE_FILE:
            if not ino["inline_len"]:
                for b in ino["direct"]:
                    if b:
                        claim(b, ino["dossier"], "data block")
                if ino["indirect"]:
                    claim(ino["indirect"], ino["dossier"], "indirect block")
                    slots = struct.unpack(f"<{img.per_index()}Q",
                                          img.block(ino["indirect"]))
                    for b in slots:
                        if b:
                            claim(b, ino["dossier"], "data block")
        else:
            problems.append(f"inode at {blk} has type {ino['type']}")

    def claim_table(node, depth):
        """The owner table's own blocks -- its index blocks and its leaves.

        Following the table's *structure* is the forward walk's business; the
        sweep below still reads only the table's contents, so the two stay
        disjoint.
        """
        if node >= total:
            problems.append(f"the owner table leaves the volume at {node}")
            return
        if by[node]:
            problems.append(f"owner table block {node} is reachable twice")
            return
        by[node] = OWNER_ARCHIVE
        if depth == 0:
            return
        slots = struct.unpack(f"<{img.per_index()}Q", img.block(node))
        for child in slots:
            if child:
                claim_table(child, depth - 1)

    # Both superblocks and the run reserved between them. Claimed explicitly,
    # rather than exempted in the comparison below -- an exemption for
    # archive-owned blocks applies to every stale copy of the root directory
    # too, and hid a leak on every commit (BG-122).
    for b in range(min(reconfs_super_b_blocks(img), total)):
        by[b] = OWNER_ARCHIVE

    claim_table(sb["table_root"], sb["table_depth"])

    walk(sb["root_inode"], 0, OWNER_ARCHIVE)

    actual = img.owners()
    for b in range(total):
        a, e = actual[b], by[b]
        if a == e:
            continue
        if a == OWNER_VOID:
            problems.append(f"block {b}: reachable but not allocated")
        elif not e:
            problems.append(f"block {b}: allocated but not reachable")
        else:
            problems.append(f"block {b}: allocated to {a}, reachable from {e}")

    root = img.inode(sb["root_inode"])
    names = [n for n, _c, _t in img.dir_entries(root)]

    found = None
    payload_round = None
    if want is not None:
        matches = [c for n, c, _t in img.dir_entries(root)
                   if n.lower() == want.lower()]
        if len(matches) == 0:
            problems.append(f"'{want}' is not in the root directory")
        elif len(matches) > 1:
            problems.append(f"'{want}' names {len(matches)} things")
        else:
            found = matches[0]
            try:
                ino = img.inode(found)
                data = img.contents(ino)

                # Only if it looks like the crash workload's payload. Other
                # callers write other things, and this reader is not only for
                # that one workload.
                if len(data) >= 4 and struct.unpack_from("<I", data, 0)[0] == \
                        PAYLOAD_MAGIC:
                    why = check_payload(data)
                    if why:
                        problems.append(f"'{want}' is not one whole version: "
                                        f"{why}")
                    else:
                        payload_round = struct.unpack_from("<I", data, 4)[0]
            except Bad as e:
                problems.append(f"'{want}' points at something unreadable: {e}")

    return {
        "epoch": sb["epoch"],
        "block_size": sb["block_size"],
        "inodes": inodes,
        "names": names,
        "target": found,
        "round": payload_round,
        "problems": problems,
    }


def damage(path, how):
    """Breaks a live structure on purpose, so a run can prove its own checker.

    Aimed through the same load() the checker uses, because the two superblocks
    alternate: a first attempt at this reached into superblock A's fields while
    B was the live one, damaged a block nothing pointed at, and the checker
    correctly reported a healthy volume. A negative control that misses its
    target reports exactly what a working checker reports.
    """
    img = Image(path)
    sb = img.load()
    data = bytearray(img.data)
    bs = img.bs

    if how == "checksum":
        # One byte inside the live root inode.
        data[sb["root_inode"] * bs + 300] ^= 0xFF
    elif how == "unallocated":
        # Say the live root inode's block belongs to nobody.
        leaf = img.leaf_block(sb["root_inode"] // img.per_leaf())
        struct.pack_into("<Q", data, leaf * bs +
                         (sb["root_inode"] % img.per_leaf()) * 8, 0)
    elif how == "torn":
        # The thing the whole crash test exists to rule out: a file made of the
        # head of one version and the tail of another.
        #
        # Built by hand rather than hoped for, because QEMU will not produce one
        # -- it does not tear a block, so this failure mode is *designed*
        # against and can only be *tested* by constructing it.
        root = img.inode(sb["root_inode"])
        target = None
        for name, child, _t in img.dir_entries(root):
            if name.lower() == "settings":
                target = child
        if target is None:
            raise Bad("no 'settings' to tear")

        ino = img.inode(target)
        if ino["inline_len"]:
            at = sb["root_inode"] * bs + INODE_SIZE + ino["size"] - 4
        else:
            # The last block holding the tail, and the round number in it.
            n = (ino["size"] - 1) // bs
            where = ino["direct"][n] if n < 12 else struct.unpack_from(
                f"<{img.per_index()}Q", img.block(ino["indirect"]))[n - 12]
            at = where * bs + ((ino["size"] - 1) % bs) - 3

        was = struct.unpack_from("<I", data, at)[0]
        struct.pack_into("<I", data, at, was + 1)
    else:
        raise Bad(f"unknown damage: {how}")

    with open(path, "wb") as f:
        f.write(data)


def main():
    if len(sys.argv) >= 3 and sys.argv[1] == "--damage":
        # reconfs-check.py --damage <how> <image>
        damage(sys.argv[3], sys.argv[2])
        print(f"damaged {sys.argv[2]}")
        return 0

    if len(sys.argv) < 2:
        print("usage: reconfs-check.py <image> [name-that-must-exist]")
        print("       reconfs-check.py --damage checksum|unallocated|torn <image>")
        return 2

    want = sys.argv[2] if len(sys.argv) > 2 else None

    try:
        r = check(sys.argv[1], want)
    except Bad as e:
        print(f"unreadable {e}")
        return 1
    except Exception as e:                    # noqa: BLE001
        print(f"unreadable {type(e).__name__}: {e}")
        return 1

    if r["problems"]:
        print(f"inconsistent epoch={r['epoch']} "
              f"{len(r['problems'])} problem(s)")
        for p in r["problems"][:5]:
            print(f"  {p}")
        return 1

    extra = f" round={r['round']}" if r["round"] is not None else ""
    print(f"ok epoch={r['epoch']} block={r['block_size']} "
          f"inodes={r['inodes']} names={','.join(r['names']) or '-'}{extra}")
    return 0


if __name__ == "__main__":
    sys.exit(main())
